pkg/jobs/jobclient.go

```diff
@@ -0,0 +1,190 @@
+					Containers: []v1.Container{
+						{
+							Name:            jobName,
+							Image:           image,
+							Command:         []string{"agent", id, script},
+							ImagePullPolicy: v1.PullAlways,
+						},
+					},
+					RestartPolicy: v1.RestartPolicyNever,
+				},
+			},
+			BackoffLimit: &backOffLimit,
+		},
+	}
+
+	job, err := jobs.Create(context.TODO(), jobSpec, metav1.CreateOptions{})
+	if err != nil {
+		log.Println("Failed to create K8s job.", err)
+	}
+
+	//print job details
+	// time.Sleep(5 * time.Second)
+
+	pods, err := c.ClientSet.CoreV1().Pods("default").List(context.TODO(), metav1.ListOptions{LabelSelector: "job-name=" + job.Name})
+	// pods, err := c.ClientSet.CoreV1().Pods("default").List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return &kubtest.ExecutionResult{
+			Status:       kubtest.ExecutionStatusError,
+			ErrorMessage: err.Error(),
+		}
+	}
+
+	for _, pod := range pods.Items {
+		if pod.Status.Phase != v1.PodRunning {
+			if pod.Labels["job-name"] == jobName {
+				if err := wait.PollImmediate(time.Second, time.Duration(0)*time.Second, isPodRunning(c.ClientSet, pod.Name, c.Namespace)); err != nil {
+					fmt.Println(err)
+					return &kubtest.ExecutionResult{
+						Status:       kubtest.ExecutionStatusError,
+						ErrorMessage: err.Error(),
+					}
+				}
+			}
+			result, err = c.GetPodLogs(pod.Name, jobName, jobName)
+			if err != nil {
+				return &kubtest.ExecutionResult{
+					Status:       kubtest.ExecutionStatusError,
+					ErrorMessage: err.Error(),
+				}
+			}
+		}
+	}
+
+	return &kubtest.ExecutionResult{
+		Status:    kubtest.ExecutionStatusSuceess,
+		RawOutput: result,
+	}
+}
+
+// connectToK8s returns ClientSet
+func connectToK8s() (*kubernetes.Clientset, error) {
+	config, err := clientcmd.BuildConfigFromFlags("", os.Getenv("KUBECONFIG"))
+	if err != nil {
+		return nil, err
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return clientset, nil
+}
+
+// isPodRunning check if the pod in question is running state
+func isPodRunning(c *kubernetes.Clientset, podName, namespace string) wait.ConditionFunc {
+	return func() (bool, error) {
+		pod, err := c.CoreV1().Pods(namespace).Get(context.Background(), podName, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+
+		switch pod.Status.Phase {
+		case v1.PodRunning, v1.PodSucceeded:
+			return true, nil
+		case v1.PodFailed:
+			return false, nil
+		}
+		return false, nil
+	}
+}
+
+func (c *JobClient) GetPodLogs(podName string, containerName string, endMessage string) (string, error) {
+	count := int64(100)
+	var toReturn string
+	var message string
+	podLogOptions := v1.PodLogOptions{
+		Follow:    true,
+		TailLines: &count,
+	}
+
+	podLogRequest := c.ClientSet.CoreV1().
+		Pods(c.Namespace).
+		GetLogs(podName, &podLogOptions)
+	stream, err := podLogRequest.Stream(context.TODO())
+	if err != nil {
+		return "", err
+	}
+
+	defer stream.Close()
+
+	for {
```

Comment:
good place to make it async - we could think about setting raw output as some kind log - and append new job output lines to it
would be also great place for integration of getting real time log output.

Reply:
log lines could be passed as channel of string ?

Reply:
Let's add more stuff to this.